### PR TITLE
Add Lunar Eclipses to Astronomical calculations

### PR DESCRIPTION
**Pull request by worachate001**
### Description

This PR partially addresses #407 by adding the method to search for lunar eclipses during user-input time span.

It shows:
- Date of time of greatest eclipse.
- Saros series
- Type of eclipses
- Penumbral and umbral magnitudes at greatest eclipse

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

### How Has This Been Tested?

**Test Configuration**:
* Operating system: Ubuntu 20.04
* Graphics Card: NVidia Geforce RTX

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
I think the main tab should be named "Eclipses" and it should contains at least 2 "inner" tabs (similar to Graphs) - "Lunar eclipses" and "Solar eclipses". Probably "Solar eclipses" subtab in future may be splitted on 2 subtabs.

**Comment by worachate001** [comment]
I got the idea, it will take some time to add solar eclipse calculations in the second tab. I think I should make it to be draft for now.

**Comment by alex-w** [comment]
> I got the idea, it will take some time to add solar eclipse calculations in the second tab. I think I should make it to be draft for now.

No need implement computation of solar eclipses right now in this branch, but prepared to compute it will be good step.

**Comment by alex-w** [comment]
Apparently the shortest way to preparation the support of both types of eclipses will be renaming the main tab to "Eclipses" at this step.

**Comment by alex-w** [comment]
I'm checked the work of tool with Astronomical Calendars and results is very good. Thank you!

So, the cosmetic fixes:
1) The phases data should be aligned to right side for improve readability and no need 4 digits for accuracy - 3 digits is enough (we use 1 digit after dot for percentage info about phases in infostring also)
2) The phase cannot be negative, so, all negative values should replaced by "dash" char (QChar(0x2014))

**Review by alex-w — APPROVED** [review]
Thank you very much!

**Comment by alex-w** [comment]
@gzotti ?

File: src/gui/AstroCalcDialog.cpp
Comment (gzotti) [inline src/gui/AstroCalcDialog.cpp:2360]: Thanks for the documentation, but please add the issue number.

File: src/gui/AstroCalcDialog.hpp
Comment (gzotti) [inline src/gui/AstroCalcDialog.hpp:165]: typo: ecipse->eclipse

File: src/gui/AstroCalcDialog.hpp
Comment (gzotti) [inline src/gui/AstroCalcDialog.hpp:217]: Please add any references here. "Algorithm taken from..." (or developed by yourself?)

**Comment by gzotti** [comment]
@worachate001 Thank you for this very welcome addition. I noticed however that the center times of this year's eclipses are off by about 1/2 minute. Are the displayed times approximate, or should they provide the displayed seconds of accuracy? Then something is not yet finished.

**Comment by worachate001** [comment]
@gzotti Thank you for the comments. Looking at the values, I think the time of greatest eclipse are accurate to within a few seconds (you have to adjust delta T for comparison and use the same ephemeris). I'll try to solve all of comments and investigate deeper.

File: src/gui/AstroCalcDialog.hpp
Comment (worachate001) [reply src/gui/AstroCalcDialog.hpp:217]: This part is just follow the same pattern from a few lines above it. I think you mean the AstroCalcDialog.cpp?

Comment (gzotti) [reply src/gui/AstroCalcDialog.hpp:217]: The header file is the place for general developer documentation (Doxygen format). Literature information (if available) should go here. Inside the cpp file of course more details can be given in further comments.

**Comment by gzotti** [comment]
> @gzotti Thank you for the comments. Looking at the values, I think the time of greatest eclipse are accurate to within a few seconds (you have to adjust delta T for comparison and use the same ephemeris). I'll try to solve all of comments and investigate deeper.

Do you give DT or zone time results? It seems we are off by 1/2 minute, but DeltaT is 72 seconds, so either you are searching by some step width and don't refine enough (then you may want to round to the nearest minute), or there is something else.

**Comment by worachate001** [comment]
Comparisons with https://eclipse.gsfc.nasa.gov/LEcat5/LE2001-2100.html

2021 May 26  11:19:53 TDT (dt=72) = 11:18:41 UT1 Stellarium = 11:18:40 UT1 
2021 Nov 19  09:04:06 TDT (dt=73) = 09:02:53 UT1 Stellarium = 09:02:52 UT1
2022 May 16  04:12:42 TDT (dt=73) = 04:11:29 UT1 Stellarium = 04:11:28 UT1
2022 Nov 08  11:00:22 TDT (dt=73) = 10:59:09 UT1 Stellarium = 10:59:08 UT1 
2023 May 05  17:24:05 TDT (dt=73) = 17:22:52 UT1 Stellarium = 17:22:50 UT1
2023 Oct 28  20:15:18 TDT (dt=74) = 20:14:04 UT1 Stellarium = 20:14:03 UT1

We are very accurate!

File: src/gui/AstroCalcDialog.hpp
Comment (worachate001) [reply src/gui/AstroCalcDialog.hpp:217]: I hope I did right here. If not, please educate me. I'm still in the learning process.

**Comment by alex-w** [comment]
@gzotti please review the changes. I got accuracy around 1 second for my location.

@worachate001 I think "Saros number" for name of second column will be more accurate term.

**Comment by worachate001** [comment]
Thanks for suggestions. While waiting for the reviews, I have been preparing the solar eclipse calculations and almost finish, it would be ready in a few days.

File: src/gui/AstroCalcDialog.cpp
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:262]: Maybe change default values for both fields will be good idea: the first field should contains first day of current year and second field - the last day of current year. This is cosmetics.

@worachate001 @gzotti how about this idea?

File: src/gui/AstroCalcDialog.cpp
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:2177]: Please add the call of this method into retranslate() - this step will solve problem of translation the column names when language is changed.

File: src/gui/AstroCalcDialog.cpp
Comment (gzotti) [reply src/gui/AstroCalcDialog.cpp:262]: Yes, indeed. A default of "search all eclipses of this year" sounds reasonable. Alternatively, from "today" to "1 year in the future".

File: src/gui/AstroCalcDialog.cpp
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:2247]: I think values `29.530588853` and `2451550.09765` should be defined as `const double` variables and just called in right places

**Comment by gzotti** [comment]
Something is still weird. Please look here and compare with the list above.
grafik
What am I doing wrong? &Delta;T algorithm is set to default.

**Comment by alex-w** [comment]
@gzotti please share your `config.ini` file

**Comment by gzotti** [comment]
[config.ini.txt](https://github.com/Stellarium/stellarium/files/7920934/config.ini.txt)
(renamed for github)

**Comment by alex-w** [comment]
> [config.ini.txt](https://github.com/Stellarium/stellarium/files/7920934/config.ini.txt)
> (renamed for github)

ΔT = 72.73 seconds with your config.ini file... WTF?!?

**Comment by Atque** [comment]
@gzotti your config says that aberration is off. Could that be the reason?

**Comment by alex-w** [comment]
> @gzotti your config says that aberration is off. Could that be the reason?

Yes, of course! After enabling aberration:
stellarium-000

**Comment by gzotti** [comment]
@Atque, many thanks! Indeed. On one of my other tasks I had switched that off and was able to overlook it...

I also just checked with WSL. (Its default config.ini). It shows your values. So I welcome this addition, @worachate001 many thanks!

**Comment by alex-w** [comment]
@worachate001 I'll merge it now and I'll add changes from my comments in the master after merge.

**Comment by worachate001** [comment]
I just woke up in the morning here, so I missed everything. Thanks for additional improvements.

**Comment by worachate001** [comment]
I have noticed that the time of greatest eclipse magnitude is always 1-2 seconds ahead of official time for every eclipses!. So it must be systematic errors -- the Moon is a little ahead of the location where it should be.

Is it possible that this may caused by the fact that aberration is not currently applied to the Moon?

**Comment by Atque** [comment]
@worachate001 It could be that Stellarium currently does not support diurnal aberration. I have noticed that lunar occultations are also off by 1 or 2 seconds when comparing to e.g. Cartes du Ciel (which takes diurnal aberration into account).

**Comment by alex-w** [comment]
Aberration is applied to the Moon now. See PR ;)

**Comment by alex-w** [comment]
see PR #2050

**Comment by worachate001** [comment]
In this window, no change of coordinates for the Moon when checked and uncheck. But for the Sun and planets, they changes. Is it normal?

aberration

**Comment by gzotti** [comment]
We have discussed that in summer. See #1626 . It seems that for the Moon aberration may be included in the ephemeris already. 

I have modified the lunar position somewhat to reflect the difference between center of figure and center of gravity. 
See EphemWrapper.cpp line 418. This explains exactly those 1-2 seconds, but should improve accuracy.
 
If diurnal aberration is noticeable in stellar occultations by the moon by more than 1 second or so, it may be worth considering. But it is not in my priority list.

**Comment by worachate001** [comment]
@gzotti Thank you for explanation. Modifying the Moon's position that way is really explain the differences. I have read that Astronomical Almanac is using this method but Fred Espenak decided to ignore it in his Catalog of Lunar Eclipses.

**Comment by gzotti** [comment]
Usually a difference of 2 seconds should not pose a problem in long-time eclipse canons, given uncertainties in &Delta;T. We would need a couple of (or rather a few thousand) securely timed (observed, not precomputed!) lunar occultations (intros and exits) to understand the limits of the current solution in Stellarium. I do not know the real figure offset between a perfect sphere and our tessellated (triangulated) polygonal Moon. See also discussion in #1959. Again, occultations are not my primary interest, but if somebody finds a solution for #1959, this would also be welcome of course. Anybody in IOTA reading this and and want to help? Of course, we have no model for the Lunar rim, so a difference with real-world conditions will likely persist. (I don't know if the displacement map can be interpreted and rendered in terms of real geometry to make the lunar rim appear jagged with the correct mountain silhouette.)

**Pull request merged**
